### PR TITLE
linux-pam: New formula: Blocked by valgrind #990

### PR DESCRIPTION
**Pull request by maxim-belkin**
- [x] Have you followed the [guidelines for contributing](https://github.com/Linuxbrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Linuxbrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

### CODE REVIEW

**Comment by maxim-belkin** [comment]
CircleCI: an error shows up because this is a new formula and `--keep-old` flag that is passed to `brew test-bot` causes an error. This behavior should be fixed as soon as Linuxbrew/homebrew-test-bot#1 is merged

File: Formula/linux-pam.rb [hunk 1]
```diff
@@ -0,0 +1,22 @@
+class LinuxPam < Formula
+  desc "Pluggable Authentication Modules for Linux"
+  homepage "http://www.linux-pam.org"
+  url "http://linux-pam.org/library/Linux-PAM-1.3.0.tar.bz2"
+  sha256 "241aed1ef522f66ed672719ecf2205ec513fd0075ed80cda8e086a5b1a01d1bb"
```
Comment (sjackman) [inline Formula/linux-pam.rb]: Add
```ruby
# tag "linuxbrew"
```

File: Formula/linux-pam.rb [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+  def install
+    ENV["XML_CATALOG_FILES"] = etc/"xml/catalog"
+
+    system "./configure", "--disable-debug",
+                          "--disable-dependency-tracking",
+                          "--disable-silent-rules",
+                          "--prefix=#{prefix}"
```
Comment (sjackman) [inline Formula/linux-pam.rb]: A small point, but I prefer the indentation style
```ruby
    system "./configure",
      "--disable-debug",
      "--disable-dependency-tracking",
      "--disable-silent-rules",
      "--prefix=#{prefix}"
```

File: Formula/linux-pam.rb [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    system "make"
+    system "make", "install"
```
Comment (sjackman) [inline Formula/linux-pam.rb]: Please add either `make check` or `make test` if it exists and runs in under 3 minutes.

File: Formula/linux-pam.rb [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+  end
+
+  test do
+    File.exist? "#{sbin}/unix_chkpwd"
```
Comment (sjackman) [inline Formula/linux-pam.rb]: I don't think this statement will do anything on its own without a call to `assert` of some kind.

File: Formula/linux-pam.rb [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    assert_match "libcrypt.so.1", shell_output("ldd #{sbin}/unix_chkpwd").chomp
```
Comment (sjackman) [inline Formula/linux-pam.rb]: I'd prefer something like
```ruby
assert_match "Usage", shell_output("#{sbin}/unix_chkpwd --help")
```

**Comment by sjackman** [comment]
Thanks for this formula, Maxim! `linux-pam` was blocking a few different formula in `core` and `science`.

File: Formula/linux-pam.rb (hunk 2 above)
Comment (maxim-belkin) [inline Formula/linux-pam.rb]: done

File: Formula/linux-pam.rb (hunk 5 above)
Comment (maxim-belkin) [inline Formula/linux-pam.rb]: added

Comment (maxim-belkin) [inline Formula/linux-pam.rb]: have to replace it with a different check.

File: Formula/linux-pam.rb (hunk 4 above)
Comment (maxim-belkin) [inline Formula/linux-pam.rb]: removed

Comment (maxim-belkin) [inline Formula/linux-pam.rb]: have to bring it back as this is the only test that worked so far

**Comment by maxim-belkin** [comment]
@sjackman formula is ready for the second round of review

File: Formula/linux-pam.rb (hunk 5 above)
Comment (maxim-belkin) [inline Formula/linux-pam.rb]: `unix_chkpwd --help` throws an error code (14). looks like all executables are not designed to be called directly.

**Comment by randy3k** [comment]
I am very excited about this PR. I can finally fix the rstudio-server linuxbrew formula.

**Comment by randy3k** [comment]
It might be more useful to pu the header files under `.../linux-pam/1.3.0/include/security`
so that the regular use
```
#include <security/pam_appl.h>
```
will still work.

**Comment by maxim-belkin** [comment]
Thanks for your feedback, @randy3k ! I've updated the PR to install headers to `security` subfolder

**Comment by maxim-belkin** [comment]
@randy3k could you please check that current formula fixes whatever issue you had with `rstudio-server` formula?

**Comment by randy3k** [comment]
@maxim-belkin it fixes the issue.

A not-well-tested formula of rstudio-server could be found here https://gist.github.com/randy3k/1d722cf7a9db4b0d131ceb012427f709

**Comment by sjackman** [comment]
Nice work, Maxim. Here's the file structure of PAM on Debian for reference: https://packages.debian.org/sid/amd64/libpam0g-dev/filelist

File: Formula/linux-pam.rb [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+      "--includedir=#{include}/security",
+      "--oldincludedir=#{include}",
+      "--enable-securedir=#{lib}/security"
+    system "make"
+    system "make", "check"
+    system "make", "install"
+  end
+
+  def post_install
+    chmod "u=rwxs,g=rx,o=rx", "#{sbin}/unix_chkpwd"
```
Comment (sjackman) [inline Formula/linux-pam.rb]: Does this command require root permissions to set the `setuid` bit on the executable?

**Comment by sjackman** [comment]
Nice work, Maxim. I'm all for merging this PR save a few minor comments above.

File: Formula/linux-pam.rb [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    ENV["XML_CATALOG_FILES"] = etc/"xml/catalog"
+
+    system "./configure",
+      "--disable-debug",
+      "--disable-dependency-tracking",
+      "--disable-silent-rules",
+      "--prefix=#{prefix}",
+      "--includedir=#{include}/security",
+      "--oldincludedir=#{include}",
+      "--enable-securedir=#{lib}/security"
```
Comment (maxim-belkin) [inline Formula/linux-pam.rb]: This is what I'm not sure about. From `./configure --help`:
```
--sysconfdir=DIR        read-only single-machine data [PREFIX/etc]
```
If we do `"--sysconfdir=#{etc}"`, we'll use  Linuxbrew's `etc` folder. Not sure if this is good or bad.

Comment (sjackman) [inline Formula/linux-pam.rb]: Yep, the Homebrew convention is to use `$HOMEBREW_PREFIX/etc` rather than `$HOMEBREW_CELLAR/FORMULA/VERSION/etc`

Comment (maxim-belkin) [inline Formula/linux-pam.rb]: ok, will add that

File: Formula/linux-pam.rb [hunk 8]
```diff
@@ -13,6 +13,7 @@ def install
     ENV["XML_CATALOG_FILES"] = etc/"xml/catalog"
 
     system "./configure",
+      "--sysconfdir=#{etc}",
```
Comment (sjackman) [inline Formula/linux-pam.rb]: Please put `--sysconfdir` last and grouped with the other `--xxdir=` options.

**Comment by sjackman** [comment]
Please squash the commits before merging.

**Comment by maxim-belkin** [comment]
### Dependencies to consider
1. `libiconv`
1. `intltool`
1. `libprelude`

From `./configure --help`:
```
  --with-libiconv-prefix[=DIR]  search for libiconv in DIR/include and DIR/lib
  --without-libiconv-prefix     don't search for libiconv in includedir and libdir
  --with-libintl-prefix[=DIR]  search for libintl in DIR/include and DIR/lib
  --without-libintl-prefix     don't search for libintl in includedir and libdir
  --with-libprelude-prefix=PFX`    Prefix where libprelude is installed (optional)
```

### Options to consider:

| Option | Explanation |
|--------|------------|
|  `--with-xml-catalog=CATALOG` | path to xml catalog to use (*instead of ENV variable*) |
| `--enable-pamlocking` | configure libpam to observe a global authentication lock |
| `--disable-prelude`       | do not use prelude |
| `--enable-read-both-confs` | read both `/etc/pam.d` and `/etc/pam.conf` files | 
| `--disable-lckpwdf` |       do not use the `lckpwdf` function |
| `--disable-cracklib` |     do not use `cracklib` |
| `--disable-audit` |        do not enable audit support |
|  `--disable-nis` |           Disable building NIS/YP support in `pam_unix` |
|  `--disable-selinux` |      do not use SELinux | 
|  `--disable-regenerate-docu` | Don't re-build documentation from XML sources |
| `--disable-nls` |          do not use Native Language Support |
| `--disable-rpath` |        do not hardcode runtime library paths |

File: Formula/linux-pam.rb (hunk 8 above)
Comment (maxim-belkin) [inline Formula/linux-pam.rb]: done

File: Formula/linux-pam.rb (hunk 6 above)
Comment (maxim-belkin) [inline Formula/linux-pam.rb]: no, because this file belongs to the current user.

File: Formula/linux-pam.rb (hunk 5 above)
Comment (maxim-belkin) [inline Formula/linux-pam.rb]:
```
$ [linuxbrew]/Cellar/linux-pam/1.3.0/sbin/mkhomedir_helper
Usage: [linuxbrew]/Cellar/linux-pam/1.3.0/sbin/mkhomedir_helper <username> [<umask> [<skeldir>]]
```
We could probably match this string exactly

**Comment by sjackman** [comment]
`iconv` is provided by `glibc` on Linuxbrew.

**Comment by sjackman** [comment]
As for options, I'd use which options are used by Linux From Scratch, and don't provide any Homebrew-style `option`.
http://www.linuxfromscratch.org/blfs/view/svn/postlfs/linux-pam.html

**Comment by sjackman** [comment]
@maxim-belkin What's the status of this PR?

**Comment by maxim-belkin** [comment]
It is ready but I'm considering adding `libprelude` formula first and then using it as a dependency here

**Comment by konstin** [comment]
Please merge this as soon as it is working. This is blocking https://github.com/Linuxbrew/brew/issues/185#issuecomment-263620486 and therefore breaks anything that depends on `util-working`.

File: Formula/linux-pam.rb [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    args = %W[
+      --disable-debug
+      --disable-dependency-tracking
+      --disable-silent-rules
+      --prefix=#{prefix}
+      --includedir=#{include}/security
+      --oldincludedir=#{include}
+      --enable-securedir=#{lib}/security
+      --sysconfdir=#{etc}
+      --with-xml-catalog=#{etc}/xml/catalog
```
Comment (maxim-belkin) [inline Formula/linux-pam.rb]: This is actually the "right" way to set xml catalog file for `linux-pam`.

**Comment by maxim-belkin** [comment]
@konstin : While this formula is likely to resolve your issue, I think you'd better first try unsetting `LD_LIBRARY_PATH` first (`unset LD_LIBRARY_PATH`) and then repeating the installation

**Comment by maxim-belkin** [comment]
Travis: `ld: cannot find -lopts`
CircleCI: `--keep-old was passed...` but otherwise success.

**Comment by maxim-belkin** [comment]
`libopts` is provided by `autogen`

**Comment by maxim-belkin** [comment]
`gnutls` is fixed, so the build should succeed.

**Comment by maxim-belkin** [comment]
failed while installing `valgrind`

**Comment by sjackman** [comment]
```
==> Installing linux-pam dependency: valgrind
```

`valgrind` is a dependency of `linux-pam` and must be bottle before this PR can be merged.
`valgrind` PR is at https://github.com/Linuxbrew/homebrew-core/pull/990
It was blocked by #1005, which was resolved by https://github.com/Linuxbrew/homebrew-core/pull/1042, but the bottle for `glibc` has not yet been updated.

**Comment by sjackman** [comment]
Blocked by `valgrind` #990

**Comment by sjackman** [comment]
Should we consider merging this PR without first updating the `glibc` bottle?

**Comment by randy3k** [comment]
what is the status of this PR? this has been sitting here for several months, I really want to submit rstudio-server linux fix, but it doesn't work without this PR.

File: Formula/linux-pam.rb (hunk 5 above)
Comment (sjackman) [inline Formula/linux-pam.rb]: Please change this line to
```ruby
assert_match "Usage", shell_output("#{sbin}/mkhomedir_helper")
```

**Comment by sjackman** [comment]
Let's ignore the `valgrind` failure, because `valgrind` is blocked by `glibc`, which isn't getting upgraded any time soon.

**Comment by maxim-belkin** [comment]
I'm fine with it residing in `linuxbrew/extra`

**Comment by maxim-belkin** [comment]
rebased on top of the new master branch

**Comment by maxim-belkin** [comment]
should I close this PR?

**Comment by sjackman** [comment]
Moved to https://github.com/Linuxbrew/homebrew-extra/pull/3

**Pull request closed without merging**